### PR TITLE
Frontend Compatibility for Multiple Promoted Addon Groups

### PR DESCRIPTION
**Pull request by chrstinalin**
Closes mozilla/addons#15226

Makes the frontend compatible with receiving a list of promoted addons rather than just one, per mozilla/addons-server#22888.

Chooses to display only the most 'important' badge.

Still allows for just one promoted addon (i.e won't break anything while mozilla/addons-server#22888 is in progress)

### CODE REVIEW

File: src/amo/components/HeroRecommendation/index.js [hunk 1]
```diff
@@ -242,7 +242,9 @@ export class HeroRecommendationBase extends React.Component<InternalProps> {
           <div className="HeroRecommendation-title-text">
             {titleText || <LoadingText width={20} />}
           </div>
-          {![LINE, RECOMMENDED].includes(promotedCategory) && !loading ? (
+          {![LINE, RECOMMENDED].some((item) =>
```
Comment (KevinMind) [inline src/amo/components/HeroRecommendation/index.js:245]: Nit: totally not a big deal in this case but I wanted to call out a few good general programming topics, especially with mapping arrays.

1) Cheap compute should go before (it was like that before) expensive compute. so if `!loading` evaluates to `false` we can skip the some/includes check. Not a big deal hear as the lists are tiny, but good to think about since theoretically promotedCategories is an arbitrarily large array.

2) negations on array methods can be hard to comprehend, and normally can be inverted.

```js
![LINE, RECOMMENDED].some((item) => promotedCategories.includes(item))
// this should be equivalent to
promotedCategories.every((cat) => [LINE, RECOMMEND].includes(cat))
```

3) nested array lookups should be avoided if possible. These are small arrays but it's a good habit to form so when it is an issue, you don't accidentally do it (I did  it last week so I've felt the pain)
4) It is good to name variables when you have sets of logical definition like in this case we are matching if the set of categories on the addon is a subset of the categories that we recommend, so you could set this to a variable and let the render body use the variable
```jsx
// probably unecessary to use a set, but is technically faster
const RECOMMENDED_CATEGORIES = Set([LINE, RECOMMENDED])
const isRecommendedAddon = !isLoading && promotedCategories.every((cat) => RECOMMENDED_CATEGORIES.has(cat));

{isRecommendedAddon ? <Foo /> : <Bar />}
```

File: src/amo/utils/addons.js [hunk 2]
```diff
@@ -112,24 +112,35 @@ export const getAddonJsonLinkedData = ({
...
   forBadging?: boolean,
-|}): PromotedCategoryType | null => {
-  let category = null;
-  if (addon && addon.promoted && addon.promoted.apps.includes(clientApp)) {
-    category = addon.promoted.category;
+|}): Array<PromotedCategoryType> => {
+  let categories: Array<PromotedCategoryType> = [];
+  const promoted = addon?.promoted;
+
+  if (promoted instanceof Array) {
```
Comment (KevinMind) [inline src/amo/utils/addons.js:127]: `Array.isArray(promoted)` is a more reliable check

File: src/amo/utils/addons.js [hunk 3]
```diff
@@ -112,24 +112,35 @@ export const getAddonJsonLinkedData = ({
...
   addon: AddonType | CollectionAddonType | SuggestionType | null | void,
   clientApp: string,
   forBadging?: boolean,
-|}): PromotedCategoryType | null => {
-  let category = null;
-  if (addon && addon.promoted && addon.promoted.apps.includes(clientApp)) {
-    category = addon.promoted.category;
+|}): Array<PromotedCategoryType> => {
+  let categories: Array<PromotedCategoryType> = [];
+  const promoted = addon?.promoted;
```
Comment (KevinMind) [inline src/amo/utils/addons.js:125]: I'm confused about the if/else statement. Why aren't we checking the prom.category truthy-ness when promoted is an array? Is that a different in behavior on the API level?

An idea of how you could prepare the array to simplify the check conditions.. assuming that diff is not expected. If we should in fact only check the category based on the value of the underlying data please add a comment explaining why that is.

```js
// return if undefined to make typechecks happy
if (!addon?.promoted) return [];


// initialize an array so the loop is simpler
const promoted = Array.isArray(addon.promoted) ? addon.promoted : [addon.promoted];
// now categories is just a filtered array of the original promoted array
// should we also check if promot.category is truthy? BTW?
let categories: Array<PromotedCategoryType> = promoted
  .filter((promo) => promo.apps.includes(clientApp));

// similar to a comment above, negated loops are confusing.
if (forBadging && categories.every((item) => !BADGE_CATEGORIES.includes(item))) {
   categories = [];
}
```

You could collapse the whole loop down to the categories variable also

```js
let categories = (
  Array.isArray(addon.promoted) ? addon.promoted : [addon.promoted]
).filter((promo) => promo.apps.includes(clientApp));
```

I think keeping the loop is good for readability later but since you now have an array of promoted categories it is very simple the rest of the logic.

File: src/amo/utils/addons.js (hunk 2 above)
Comment (KevinMind) [reply src/amo/utils/addons.js:127]: See other comment.

File: src/amo/components/AddonBadges/index.js [hunk 4]
```diff
@@ -35,27 +35,31 @@ type InternalProps = {|
...
 
-    const promotedCategory = _getPromotedCategory({
+    const promotedCategories = _getPromotedCategories({
       addon,
       clientApp,
       forBadging: true,
     });
 
     return (
       <div className="AddonBadges">
```
Comment (KevinMind) [inline src/amo/components/AddonBadges/index.js:55]: Can you add a screenshot with the maximum realistic number of categories.. based on addons-server can we have 1,2,3,20 categories? How would that look in the UI? I'm surprised to see 0 css changes given this type of patch.

**Review by KevinMind — DISMISSED** [review]
I would approve this, with some nits, and verify with the addons-server PR after this is merged. wdyt?

File: src/amo/utils/addons.js (hunk 3 above)
Comment (chrstinalin) [reply src/amo/utils/addons.js:125]: > I'm confused about the if/else statement. Why aren't we checking the prom.category truthy-ness when promoted is an array? Is that a different in behavior on the API level?

Typescript is specifically expecting addons.promoted to be an array of type Array<PromotedType>, so it being an Array -> assumes promo has the category. Which was different from the else section, since it could be of type Promoted | null. I guess we could do the check if we want to suppress that runtime error? But I don't see any reason to given that category not existing in the Array item would, well, **be** an error.

It's also why the filter doesn't work, since we're returning an array of PromotedCategoryType, but the promoted const is of type PromotedType. Everything else makes sense though.

File: src/amo/components/AddonBadges/index.js (hunk 4 above)
Comment (chrstinalin) [reply src/amo/components/AddonBadges/index.js:55]: There's 4 that show up as a badge -- Line, recommended, spotlight, and strategic.

image

Though, now that I'm looking at this, I'm actually not sure if those blacked out recommended badges (spotlight and strategic) are meant to be revealed.

I don't recall (off the top of my head, at least) any existing checks to avoid Spotlight or Strategic badges (both from addons-server passing them as categories, nor from frontend), and there's also this mozilla/addons#14987 issue that seems to relate to this... @diox could you shed some light on this?

>I'm surprised to see 0 css changes given this type of patch.

The AddonBadges div also already holds multiple badges (not exclusively for promoted, ex. the "Available on Firefox for Android" badge), which is why its able to handle this change without any css edits. Whether it actually looks good in the case where every single badge is available, well...

image

Comment (KevinMind) [reply src/amo/components/AddonBadges/index.js:55]: Right, the second case is what I was after. From a UX perspective, what is the maximum amount of content, and does it look good. Let's ask Nicole and yes, we should either add svgs for the blacked out badges, or filter them.

Also, why do they all say recommended? I would expect we should name them appropriately as well..

File: tests/unit/amo/utils/test_addons.js [hunk 5]
```diff
@@ -183,50 +183,157 @@ describe(__filename, () => {
...
   });
 
-  describe('getPromotedCategory', () => {
-    it('returns null if the addon is not promoted', () => {
+  describe('getPromotedCategories', () => {
+    it('returns the empty list if the addon is not promoted', () => {
       const addon = createInternalAddonWithLang({
         ...fakeAddon,
-        promoted: null,
+        promoted: [],
```
Comment (willdurand) [inline tests/unit/amo/utils/test_addons.js:190]: You need to keep this test with `null` for backward compatibility.

File: tests/unit/amo/utils/test_addons.js [hunk 6]
```diff
@@ -183,50 +183,157 @@ describe(__filename, () => {
...
-      ).toEqual(null);
+      ).toEqual([]);
     });
 
-    it('returns null if the addon is not promoted for the specified app', () => {
+    it('returns the empty list if the addon is not promoted for the specified app', () => {
       const addon = createInternalAddonWithLang({
         ...fakeAddon,
-        promoted: { category: RECOMMENDED, apps: [CLIENT_APP_ANDROID] },
+        promoted: [{ category: RECOMMENDED, apps: [CLIENT_APP_ANDROID] }],
```
Comment (willdurand) [inline tests/unit/amo/utils/test_addons.js:210]: Same here, we should keep a test with `promoted` as an object (and not a list)

**Review by willdurand — DISMISSED** [review]
That's a good start, thanks. I provided some diffs to help you get to where I'd like us to go.

Additionally, I noticed a bunch of UI glitches:

- For instance, how should we treat themes?
Screenshot 2024-12-09 at 17 19 35

- Badges in the search results should have some padding in between:
Screenshot 2024-12-09 at 17 32 56

- Multiple badges in the "more add-ons by this developer" card look a bit off:
Screenshot 2024-12-09 at 17 35 19

- same for "more themes by artist":
Screenshot 2024-12-09 at 17 36 20

- and in the user profile:


- and in the "other popular add-ons" card:
Screenshot 2024-12-09 at 17 37 21

File: tests/unit/amo/utils/test_addons.js (hunk 5 above)
Comment (chrstinalin) [reply tests/unit/amo/utils/test_addons.js:190]: I had moved it to a different test on line 316

File: tests/unit/amo/utils/test_addons.js (hunk 6 above)
Comment (chrstinalin) [reply tests/unit/amo/utils/test_addons.js:210]: Also in a different test, line 336

**Comment by KevinMind** [comment]
@willdurand idk what is happening but github refuses to let me remove myself, in the process of wrestling the UI I re-requested review from you (accidentally)

File: src/amo/reducers/autocomplete.js [hunk 7]
```diff
@@ -27,7 +30,7 @@ export type SuggestionType = {|
   addonId: number,
   iconUrl: string,
   name: string,
-  promoted: PromotedType | null,
+  promoted: Array<PromotedType> | null,
```
Comment (willdurand) [inline src/amo/reducers/autocomplete.js:33]:
```suggestion
  promoted: Array<PromotedType>,
```

File: src/amo/utils/addons.js [hunk 8]
```diff
@@ -121,15 +122,23 @@ export const getPromotedCategory = ({
   clientApp: string,
   forBadging?: boolean,
 |}): PromotedCategoryType | null => {
-  let category = null;
-  if (addon && addon.promoted && addon.promoted.apps.includes(clientApp)) {
-    category = addon.promoted.category;
-  }
+  if (!addon?.promoted) return null;
```
Comment (willdurand) [inline src/amo/utils/addons.js:125]:
```suggestion
  if (addon.promoted.length === 0) {
    return null;
  }
```

File: src/amo/utils/addons.js [hunk 9]
```diff
@@ -121,15 +122,23 @@ export const getPromotedCategory = ({
...
+    })
+    .map((promoted) => promoted.category)
+    .sort(
+      (a, b) =>
+        ALL_PROMOTED_CATEGORIES.indexOf(a) - ALL_PROMOTED_CATEGORIES.indexOf(b),
+    );
 
-  return category;
+  // Return only the 'most important' badge.
+  return categories.length > 0 ? categories[0] : null;
```
Comment (willdurand) [inline src/amo/utils/addons.js:143]:
```suggestion
  return categories.shift() || null;
```

File: tests/unit/amo/pages/TestAddon.js [hunk 10]
```diff
@@ -2965,6 +2968,18 @@ describe(__filename, () => {
...
+    it('does not render the strategic or spotlight badges and correctly renders only the most important badge (RECOMMENDED)', () => {
+      const categories = [LINE, RECOMMENDED, STRATEGIC, SPOTLIGHT];
+      addon.promoted = categories.map((category) => ({
+        category,
+        apps: [clientApp],
+      }));
+      renderWithAddon();
+      expect(screen.getByClassName('PromotedBadge')).toHaveClass(
+        `PromotedBadge--recommended`,
+      );
```
Comment (willdurand) [inline tests/unit/amo/pages/TestAddon.js:2980]: This doesn't tell me that other badges aren't shown, maybe make sure that the badge for line extensions isn't in the document.

File: tests/unit/amo/utils/test_addons.js [hunk 11]
```diff
@@ -249,6 +249,80 @@ describe(__filename, () => {
...
+      ).toEqual(null);
+      expect(
+        getPromotedCategory({
+          addon: suggestion,
+          clientApp: CLIENT_APP_ANDROID,
+        }),
+      ).toEqual(null);
+    });
+
+    it('returns the category if the addon is promoted for the specified app if passed PromotedType directly', () => {
```
Comment (willdurand) [inline tests/unit/amo/utils/test_addons.js:303]:
```suggestion
    it('returns the category if the addon is promoted for the specified app', () => {
```

File: tests/unit/amo/utils/test_addons.js [hunk 12]
```diff
@@ -249,6 +249,80 @@ describe(__filename, () => {
...
+      ).toEqual(categories[0]);
+      expect(
+        getPromotedCategory({
+          addon: suggestion,
+          clientApp: CLIENT_APP_ANDROID,
+        }),
+      ).toEqual(categories[0]);
+    });
+
+    it('returns null if the addon is not promoted via null', () => {
```
Comment (willdurand) [inline tests/unit/amo/utils/test_addons.js:283]: Maybe add a test with `promoted: []` too

File: src/amo/reducers/utils.js [hunk 13]
```diff
@@ -26,3 +26,12 @@ export const selectCategoryObject = (
 ): CategoryEntry => {
   return apiAddon.categories;
 };
+
+export const makeInternalPromoted = (
```
Comment (willdurand) [inline src/amo/reducers/utils.js:30]: We should probably have some test coverage for this function.

**Review by willdurand — CHANGES_REQUESTED** [review]

File: src/amo/utils/addons.js (hunk 8 above)
Comment (chrstinalin) [reply src/amo/utils/addons.js:125]: The definition of getPromotedCategory has
`addon: AddonType | CollectionAddonType | SuggestionType | null | void,`
This makes Typescript mad

Comment (willdurand) [reply src/amo/utils/addons.js:125]: ha, I see

**Review by willdurand — APPROVED** [review]
